server/schedule/checker/rule_checker.go

```diff
@@ -0,0 +1,266 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checker
+
+import (
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/log"
+	"github.com/pingcap/pd/server/core"
+	"github.com/pingcap/pd/server/schedule/filter"
+	"github.com/pingcap/pd/server/schedule/operator"
+	"github.com/pingcap/pd/server/schedule/opt"
+	"github.com/pingcap/pd/server/schedule/placement"
+	"github.com/pingcap/pd/server/schedule/selector"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+// RuleChecker fix/improve region by placement rules.
+type RuleChecker struct {
+	cluster     opt.Cluster
+	ruleManager *placement.RuleManager
+	name        string
+}
+
+// NewRuleChecker creates an checker instance.
+func NewRuleChecker(cluster opt.Cluster, ruleManager *placement.RuleManager) *RuleChecker {
+	return &RuleChecker{
+		cluster:     cluster,
+		ruleManager: ruleManager,
+		name:        "rule-checker",
+	}
+}
+
+// Check checks if the region matches placement rules and returns Operator to
+// fix it.
+func (c *RuleChecker) Check(region *core.RegionInfo) *operator.Operator {
+	checkerCounter.WithLabelValues("rule_checker", "check").Inc()
+
+	fit := c.cluster.FitRegion(region)
+	if len(fit.RuleFits) == 0 {
+		// If the region matches no rules, the most possible reason is it spans across
+		// multiple rules.
+		return c.fixRange(region)
+	}
+	for _, rf := range fit.RuleFits {
+		op, err := c.fixRulePeer(region, fit, rf)
+		if err != nil {
+			log.Debug("fail to fix rule peer", zap.Error(err), zap.String("rule-group", rf.Rule.GroupID), zap.String("rule-id", rf.Rule.ID))
+			return nil
+		}
+		if op != nil {
+			return op
+		}
+	}
+	op, err := c.fixOrphanPeers(region, fit)
+	if err != nil {
+		log.Debug("fail to fix orphan peer", zap.Error(err))
+		return nil
+	}
+	return op
+}
+
+func (c *RuleChecker) fixRange(region *core.RegionInfo) *operator.Operator {
+	keys := c.ruleManager.GetSplitKeys(region.GetStartKey(), region.GetEndKey())
+	if len(keys) == 0 {
+		return nil
+	}
+	return operator.CreateSplitRegionOperator("rule-split-region", region, 0, pdpb.CheckPolicy_USEKEY, keys)
+}
+
+func (c *RuleChecker) fixRulePeer(region *core.RegionInfo, fit *placement.RegionFit, rf *placement.RuleFit) (*operator.Operator, error) {
+	// make up peers.
+	if len(rf.Peers) < rf.Rule.Count {
+		return c.addRulePeer(region, rf)
+	}
+	// fix down/offline peers.
+	for _, peer := range rf.Peers {
+		if c.isDownPeer(region, peer) || c.isOfflinePeer(region, peer) {
+			return c.replaceRulePeer(region, fit, rf, peer)
```

Comment:
Do we need to distinguish these two types of operators?

Reply:
fixed.